[jsonpath] fix wrong output when using jsonpath

zerodayz opened the pull request:
/kind bug

**Which issue(s) this PR fixes**:
Fix range loop when using jsonpath

Without patch:

```
kubectl get -n openshift-oauth-apiserver po -o jsonpath='{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}'
apiserver-7d9cc97649-79c2x is Running
apiserver-7d9cc97649-lgks6 is Running
apiserver-7d9cc97649-qgkxn is Running
 is Running
```

With patch:

```
kubectl get -n openshift-oauth-apiserver po -o jsonpath='{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}'
apiserver-7d9cc97649-79c2x is Running
apiserver-7d9cc97649-lgks6 is Running
apiserver-7d9cc97649-qgkxn is Running
```

Fixes BZ#1900755

zerodayz committed:
[jsonpath] fix wrong output when using jsonpath

Fix range loop when using jsonpath

Without patch:

kubectl get -n openshift-oauth-apiserver po -o jsonpath='{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}'
apiserver-7d9cc97649-79c2x is Running
apiserver-7d9cc97649-lgks6 is Running
apiserver-7d9cc97649-qgkxn is Running
 is Running

With patch:

kubectl get -n openshift-oauth-apiserver po -o jsonpath='{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}'
apiserver-7d9cc97649-79c2x is Running
apiserver-7d9cc97649-lgks6 is Running
apiserver-7d9cc97649-qgkxn is Running

soltysh commented:
/ok-to-test

fedebongio commented:
/assign @caesarxuchao 
/triage accepted

caesarxuchao commented:
Thanks for contributing, @zerodayz.

This broke existing unit tests: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/98057/pull-kubernetes-bazel-test/1349680278016102400

Also can you add a unit test for the problem you are fixing?

zerodayz commented:
@caesarxuchao Good morning :) I will take a look what happened there and also add Unit test for this one. Cheers!

zerodayz commented:
I can see what happened, the new line didn't get appended since I broke out from the loop. And that continue ensure the new line is there. It was masked for me because my use case :  '{" is Running\n"}'  which added new line for me instead, so output was alright. I am scratching my first patch and will start with Unit Test. Thanks @caesarxuchao

zerodayz commented:
Added UnitTest to show what I am testing, but the UnitTest doesn't reflect the reality from what we see on kubectl. I am confused a bit.

https://github.com/kubernetes/kubernetes/pull/98057/files < UnitTest

As you can see we are ranging thru items and the last one is empty, in kubectl it prints out extra line:

```
kubectl get -n openshift-oauth-apiserver po -o jsonpath='{range .items[?(.status.phase=="Running")]}{.metadata.name}{" is Running\n"}'
apiserver-7d9cc97649-79c2x is Running
apiserver-7d9cc97649-lgks6 is Running
apiserver-7d9cc97649-qgkxn is Running
 is Running
```

And this is because the items contains `{"resourceVersion":"","selfLink":""}` at the end of the list, I added it to the UnitTest but it just wouldn't print that .

zerodayz commented:
Forget what I just said, I noticed this fails with 3 and more pods. I updated unit test and now its properly failing.

zerodayz commented:
Now UnitTest is properly there, I am looking into the problem and will test locally. Sorry for the noise.

zerodayz commented:
/retest

zerodayz commented:
/retest

zerodayz commented:
Hi @caesarxuchao  I believe that problem is fixed right now, do you happen to have time to check? I am also 👀 @soltysh  as he authored the initial code.

RainbowMango commented:
/assign @caesarxuchao
I've met this issue at https://github.com/kubernetes-sigs/mcs-api/issues/3.

zerodayz commented:
I also hit this in the test env when I am preparing for the script to collect APF endpoints.

zerodayz commented:
/cc @caesarxuchao Would you be able to review please? Thank you :-)

soltysh reviewed (approved):
/lgtm

ohbus reviewed (approved):
/lgtm

zerodayz commented:
@deads2k or @yliaog Any comments guys? It has been waiting for review for almost month. Do you think you will find some time in the near future? Thank you!

yliaog commented:
/assign @caesarxuchao

dims commented:
/release-note-none
/assign @liggitt @sttts

caesarxuchao commented:
/approve
/lgtm

sorry for the long delay.

pacoxu commented:
/test pull-kubernetes-unit

soltysh commented:
@zerodayz it looks like this PR is breaking backwards compatibility, in the following scenario:

```
$ kubectl get po --template='bad{ .missingkey }key' --output=jsonpath
bad
```
the output is wrong, in previous versions you'd get:
```
$ kubectl get po --template='bad{ .missingkey }key' --output=jsonpath
badkey
```
only passing `--allow-missing-template-keys=false` would error that last one out.
